src/main/resources/webapp/interface/dynamic/addDynamicTypes.es

```diff
@@ -0,0 +1,201 @@
+import {
+	VALUE_TYPE_BOOLEAN,
+	VALUE_TYPE_DOUBLE,
+	VALUE_TYPE_GEO_POINT,
+	VALUE_TYPE_INSTANT,
+	VALUE_TYPE_LOCAL_DATE,
+	VALUE_TYPE_LOCAL_DATE_TIME,
+	VALUE_TYPE_LOCAL_TIME,
+	VALUE_TYPE_LONG,
+	VALUE_TYPE_REFERENCE,
+	VALUE_TYPE_STRING,
+	camelize,
+	forceArray,
+	sortKeys,
+	toStr,
+	ucFirst
+} from '@enonic/js-utils';
+import setIn from 'set-value'; // Number.isInteger and Reflect
+
+import {GraphQLString} from '/lib/graphql';
+
+import {
+	GQL_INPUT_FIELDS_HIGHLIGHT_PROPERTIES,
+	GQL_INTERFACE_TYPE_DOCUMENT,
+	//GQL_OBJECT_TYPE_GLOBAL_FIELD,
+	GQL_OBJECT_TYPE_INTERFACE_SEARCH_HIT_HIGHLIGHT
+} from '../constants';
+import {documentTypeNameToGraphQLObjectTypeName} from './documentTypeNameToGraphQLObjectTypeName';
+import {objToGraphQL} from './objToGraphQL';
+import {valueTypeToGraphQLType} from './valueTypeToGraphQLType';
+
+import {addDynamicEnumTypes} from './addDynamicEnumTypes';
+import {addDynamicInputTypes} from './addDynamicInputTypes';
+import {addDynamicInterfaceTypes} from './addDynamicInterfaceTypes';
+import {addDynamicObjectTypes} from './addDynamicObjectTypes';
+//import {addDynamicUnionTypes} from './addDynamicUnionTypes';
+
+
+const VALUE_TYPE_VARIANTS = [
+	VALUE_TYPE_BOOLEAN,
+	VALUE_TYPE_DOUBLE,
+	VALUE_TYPE_GEO_POINT,
+	VALUE_TYPE_INSTANT,
+	VALUE_TYPE_LOCAL_DATE,
+	VALUE_TYPE_LOCAL_DATE_TIME,
+	VALUE_TYPE_LOCAL_TIME,
+	VALUE_TYPE_LONG,
+	VALUE_TYPE_REFERENCE,
+	VALUE_TYPE_STRING
+];
+
+/*──────────────────────────────────────────────────────────────────────────────
+An objectType per documentType will be generated.
+Global and "magic" fields will be added to an interfaceType.
+The list of objectTypes is used both in the interfaceType and unionType resolvers.
+When we define the resolvers we may use a js-reference to an empty object, which later will contain the objectTypes.
+When we define the unionType, we can use lib-graphql.reference, but we need to have a list of the objectType names, so we might as well use the objectTypes directly.
+Q: Can we use lib-graphql.reference in objectType.interfaces?
+──────────────────────────────────────────────────────────────────────────────*/
+
+export function addDynamicTypes({
+	allFieldKeys,
+	camelToFieldObj,
+	documentTypes,
+	globalFieldsObj,
+	glue,
+	interfaceSearchHitsHighlightsFields
+}) {
+	const documentTypeObjectTypes = {}; // Defined before addDynamicInterfaceTypes, populated after
+
+	// Must be before addDynamicObjectTypes
+	addDynamicInterfaceTypes({
+		documentTypeObjectTypes, // Just an empty obj, populated later
+		glue,
+		interfaceSearchHitsHighlightsFields,
+		globalFieldsObj
+	});
+
+	const interfaceTypeDocument = glue.getInterfaceType(GQL_INTERFACE_TYPE_DOCUMENT);
+	documentTypes.forEach(({
+		_name: documentTypeName,
+		properties
+	}) => {
```

Comment:
Should be extracted to a separate function with explanatory name.

Reply:
Yes